### ISSUE TITLE
useDataStoreBinding results in infinite useEffect loop when criteria is set with createDataStorePredicate

### ISSUE THREAD

**Issue by dpilch** [aws-amplify/amplify-ui#1396]
### Before creating a new issue, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-ui/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-ui/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md#bug-reports).

### On which framework/platform are you having an issue?

React

### Which UI component?

Other

### How is your app built?

Create React App

### Please describe your bug.

With Studio component when a predicate is added to a `useDataStoreBinding` the hook gets stuck in an infinite loop.

Warning from console: `Warning: Maximum update depth exceeded. This can happen when a component calls setState inside useEffect, but useEffect either doesn't have a dependency array, or one of the dependencies changes on every render.`

### What's the expected behaviour?

`createDataStorePredicate` should use `useCallback` with a dependency on inputted object so that a new function is not created each render.

This does result in another issue because in the example code snippet `buttonUserFilterObj` is redefined on each render. There are a few solutions to this.
1. Add custom comparison logic to the useCallback by stringifying the object.
1. define the object with useMemo (this would be difficult for codegen because of the recursive nature of the filter object).
1. define the object with useState.
    1. this could be done in the component or within the createDataStorePredicate function.

```typescript
const [buttonUserFilterObj] = React.useState({
  field: "firstName",
  operand: "Dane",
  operator: "eq",
})

// or 

export const createDataStorePredicateCallback = <Model extends PersistentModel>(
  predicateObject: DataStorePredicateObject
): ProducerModelPredicate<Model> => {
  const [predicateObjectState] = React.useState(predicateObject);
...
```

I think the option 3 is easiest to implement and least error prone.

I tried option with this quick solution and it resolved the issue.
```typescript
const createDataStorePredicate = (predicateObject) => {
  const [predicateObjectState] = React.useState(predicateObject);
  return React.useCallback((...args) => {
    return createDataStorePredicate(predicateObjectState)(...args)
  }, [predicateObjectState]);
}
```

### Help us reproduce the bug!

Follow the integration setup on [Amplify Codegen UI tagged-release/q1-release](https://github.com/aws-amplify/amplify-codegen-ui/tree/tagged-release/q1-release).

### Code Snippet

```typescript
const buttonUserFilterObj = {
  field: "firstName",
  operand: "Dane",
  operator: "eq",
};

// creates a new predicate function each time the component renders
const buttonUserFilter = createDataStorePredicate(buttonUserFilterObj);

// this hook will cause infinite loop
const res = useDataStoreCollection({
  type: "collection",
  model: User,
  // the useEffect in useDataStoreCollection has a dependency on criteria
  // however, criteria is a new function on each render so it is always considered not equal and the hook runs
  criteria: buttonUserFilter,
});
```

### Additional information and screenshots

This error only happens on the tagged-release/q1-release branch of amplify-codegen-react and version 2.6.1 of Amplify UI React. It is unclear how this is currently working on other versions. amplify-codegen-react and Amplify UI did not appear to make changes to code that would affect this feature.

**Comment by dpilch** [comment]
We found that as part of the React import refactor dependencies were added to the useEffect hooks. https://github.com/aws-amplify/amplify-ui/commit/43c814db418865672730449ba8c983bcb7d94847#diff-78fd8656c9b68a3d3d93c93d975912ac7b5c8751beb3d7885b1df7f25cc911b6R74

This additional change was initially missed.

With the dependencies set as an empty array the hooks will run only once (preventing the infinite useEffect loop). This change will be fully reverted to restore functionality.

The dependencies will need to be added to the hooks, but will probably need to be added as a breaking change.

**Comment by zchenwei** [comment]
I had a discussion with Scott offline. Having an empty dependencies array is by design. Only running useEffect once inside the useDataStore hook is an expected behavior because we do not expect the parameters would change dynamically. I have raised a PR to add some comments and disable the eslint rules there. Thanks!
